File: 15-Java8/src/by/gsu/epamlab/beans/ExtraTrial.java
package by.gsu.epamlab.beans;

import static by.Constants.*;

public class ExtraTrial extends Trial{
    private int mark3;
    private static final int THIRD_EXAM_PASS_MARK = 60;

    public ExtraTrial() {
    }

    public ExtraTrial(String name, int mark1, int mark2, int mark3) {
        super(name, mark1, mark2);
        setMark3(mark3);
    }

    public int getMark3() {
        return mark3;
    }

    public void setMark3(int mark3) {
        if (mark3 > MARK_MAX_VALUE || mark3 < MARK_MIN_VALUE){
            throw new IllegalArgumentException(EXC_ILLEGAL_MARK3);
        }
        this.mark3 = mark3;
    }

    @Override
    public void clearAllResults(){
        super.clearAllResults();
        mark3 = MARK_MIN_VALUE;
    }

    @Override
    public boolean isPassed() {
        return super.isPassed() && getMark3() >= THIRD_EXAM_PASS_MARK;
    }

    @Override
    public Trial getCopy(){
        return new ExtraTrial(getName(), getMark1(), getMark2(), mark3);
    }

    @Override
    protected String fieldsToString(){
        return super.fieldsToString() + DELIMITER + mark3;
    }

}

File: 22-jsOop/beans/ExtraTrial.js
const Trial = require("./Trial.js");

class ExtraTrial extends Trial{
    static THIRD_EXAM_PASS_MARK = 60;

    constructor(name, mark1, mark2, mark3) {
        super(name, mark1, mark2);
        if(mark3 > Trial.MARK_MAX_VALUE || mark3 < Trial.MARK_MIN_VALUE){
            throw new RangeError("Illegal argument in mark3");
        }
        this._mark3 = mark3;
    }

    get mark3(){
        return this._mark3;
    }

    get clearAllResults() {
        this._mark1 = 0;
        this._mark2 = 0;
        this._mark3 = 0;
        return this;
    }

    isPassed() {
        return super.isPassed() && this._mark3 >= ExtraTrial.THIRD_EXAM_PASS_MARK;
    }

    getCopy(){
        return new ExtraTrial(this._account, this._mark1, this._mark2, this._mark3);
    }

    fieldsToString() {
        return `${super.fieldsToString()};${this._mark3}`;
    }
}
module.exports = ExtraTrial;